This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [2]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case


Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [3]:
# First, import the relevant modules
import requests


In [25]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url_sample = "https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2014-01-01&end_date=2014-01-01&api_key=" + API_KEY
r_sample = requests.get(url_sample)

In [26]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
print(r_sample.json())

{'dataset_data': {'limit': None, 'transform': None, 'column_index': None, 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'start_date': '2014-01-01', 'end_date': '2014-01-01', 'frequency': 'daily', 'data': [], 'collapse': None, 'order': None}}


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [14]:
url_2017 = "https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2017-01-01&end_date=2017-12-31&api_key=" + API_KEY

In [28]:
r_2017 = requests.get(url_2017)
data_dict = r_2017.json()

In [35]:
  #printing column names to know what indices will be needed to answer the 
  #above questions
print(data_dict['dataset_data']['column_names'])

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


In [52]:
data = data_dict['dataset_data']['data']

In [72]:
open_price = [(x[1], x[0]) for x in data if x[1] != None]

In [73]:
max(open_price), min(open_price)

((53.11, '2017-12-14'), (34.0, '2017-01-24'))

**Question 3** 
The highest opening price was \\$53.11 on 12/14 and the lowest was \\$34.00 on   1/24.

In [75]:
price_diff = [(x[2] - x[3], x[0]) for x in data]

In [78]:
max(price_diff)

(2.8100000000000023, '2017-05-11')

**Question 4** The largest change between high and low price in a given day was \\$2.81 and occurred on 5/11.

In [89]:
price_change = [(abs(x[5]), x[0], x[5]) for x in data if x[5] != None]

In [90]:
max(price_change)

(0.44, '2017-05-01', -0.44)

**Question 5** The largest change between days was a decrease of \$$0.44 and occurred 5/1.

In [114]:
trade_volume = [x[6] for x in data if x[6] != None]

In [115]:
type(trade_volume)

list

In [116]:
sum(trade_volume) / len(trade_volume)

89124.33725490196

**Question 6** Average daily trading volume was 89124.

In [119]:
trade_volume.sort()

In [123]:
len(trade_volume)

255

In [124]:
trade_volume[122]

73322.0

**Question 7** Median trade volume was 73322!